src/libstd/thread/local.rs

```diff
@@ -97,10 +96,19 @@ pub struct LocalKey<T: 'static> {
     // trivially devirtualizable by LLVM because the value of `inner` never
     // changes and the constant should be readonly within a crate. This mainly
     // only runs into problems when TLS statics are exported across crates.
-    inner: unsafe fn() -> Option<&'static UnsafeCell<Option<T>>>,
 
-    // initialization routine to invoke to create a value
+    // The user-supplied initialization function that constructs a new T.
     init: fn() -> T,
+
+    // Platform-specific callbacks.
+
+    // Get a reference to the value. get can be called at any time during the
+    // value's life cycle (Uninitialized, Initializing, Valid, and Destroyed).
+    get: unsafe fn() -> &'static UnsafeCell<LocalKeyValue<T>>,
+
+    // Register a destructor for the value. register_dtor should be called
+    // after the key has been transitioned into the Valid state.
+    register_dtor: unsafe fn(),
```

Comment:
Couldn't this still happen on the first call to `get`? If initialization fails that means we'll just run the destructor on a noop value, right?

Reply:
I wanted to avoid adding an extra branch in the hot path, and that would require branching in `get` to see whether the destructor had been registered already.